### ISSUE TITLE
enterprise restore failure

### ISSUE THREAD

**Issue by zhangquan913** [cockroachdb/cockroach#34038]
**Describe the problem**

After getting some response from https://github.com/cockroachdb/cockroach/issues/34031, I tried to rerun the incremental backup process again in crdb v2.0.3. This time the error is even weirder. The error occurred even before the incremental backup step.

The database is still “test_db”, which contains two tables “backup_test” and “config”. “backup_test” contains 6 records and “config” contains 1 record.

**To Reproduce**

```
root@:26257/test_db> backup database “test_db” to ‘nodelocal:///20190115-1/full_backup’;
±-------------------±----------±-------------------±-----±--------------±---------------±------+
| job_id | status | fraction_completed | rows | index_entries | system_records | bytes |
±-------------------±----------±-------------------±-----±--------------±---------------±------+
| 417789178037436417 | succeeded | 1 | 7 | 12 | 0 | 1145 |
±-------------------±----------±-------------------±-----±--------------±---------------±------+
(1 row)

Time: 250.759485ms
```

```
root@:26257/test_db> drop database “test_db” cascade;
DROP DATABASE

Time: 220.445184ms

warning: error retrieving the database name: pq: database “test_db” does not exist

```
```
root@:26257> show databases;
±---------+
| Database |
±---------+
| system |
±---------+
(1 row)

Time: 3.173739ms

warning: error retrieving the database name: pq: database “test_db” does not exist

```
```
root@:26257> restore database “test_db” from ‘nodelocal:///20190115-1/full_backup’;
pq: importing 4 ranges: fetching “417789178255540225.sst”: open /mnt/ssd-0/extern/20190115-1/full_backup/417789178255540225.sst: no such file or directory
warning: error retrieving the database name: pq: database “test-db” does not exist

```
It kept showing warning “test_db” does not exist after dropping db. However, after executing the drop command, db should not exist as expected. Why there are warning here. Is this a bug crdb should improve? And for the missing file or directory message, I am sure the file was there with the right access permission.

**Expected behavior**
1. Warning of "database does not exist" should not show after a drop database command.
2. Resolve the weird .sst file missing message.

**Comment by knz** [comment]
1) The warning `warning: error retrieving the database name` is innocuous and irrelevant to your issue.

2) the error `open /mnt/ssd-0/extern/20190115-1/full_backup/417789178255540225.sst: no such file or directory` suggests that you have used `nodelocal://` with a multi-node cluster. Is that possible?

**Comment by knz** [comment]
also we have corrected a lot of bugs in the 2.0.x series (and more towards 2.1.x). If you're using 2.0.3 you are likely encountering several of them. Check all the release notes since 2.0.3 to find backup/restore fixes.

**Comment by zhangquan913** [comment]
@knz nodelocal cannot be used for a multi-node cluster? CRDB is always deployed on the multi-node cluster. It is ridiculous if it only supports a single node. What should be used for multi-node cluster if I want to back up to the local file system on that cluster?

**Comment by knz** [comment]
@zhangquan913 you can use `nodelocal://` if you can **GUARANTEE** that all the nodes are using a shared, highly consistent network filesystem for their `--external-io` path. (Hint: NFS is *not* such a filesystem).

If you do not use a shared, highly-consistent network filesystem, the BACKUP mechanism will create corrupt backups which you cannot restore.

**This is extensively documented already.**

How would you like us to help further? Please advise what you would like to us to do.

**Comment by knz** [comment]
@tim-o please accompany this issue further thanks

**Comment by knz** [comment]
If you can fulfill the requirements, that is a good starting point for a discussion about your observations!

That said, we cannot honestly guarantee anything about 2.0.3 because this version is simply too old and not supported any more.

There is *some* guarantee available with the latest 2.0.x version (at least, we will be willing to help you diagnose further), but if you find any real bug we will only fix these bugs in the next 2.1.x version or 2.2.

**Comment by zhangquan913** [comment]
@knz Thanks so much today. Let me talk to my team tomorrow to know if we fulfilled the requirements and further discuss if we have to use the 2.1 version.

**Comment by zhangquan913** [comment]
@knz I reviewed all the release notes since 2.0.3. Did not find any bug fix related to the issues I posted. Just curious, was the "retrieving db name warning" and ".sst file missing" issues the first time raised by users?

**Comment by knz** [comment]
1) The "retrieving db name" warning is a warning issued by `cockroach sql` because you asked it to connect to a db named `test_db` and you are DROPping it from "under it". It's simply complaining (as it should!) that the database you want it to find, isn't there. There is nothing wrong about that and it is a natural consequence of your actions. It's also, as I wrote above, not relevant to your backup/restore problem.

2) the ".sst file missing" is a symptom of BACKUP writing the files in one filesystem, and RESTORE trying to find them in a different filesystem (presumably because the distributed RESTORE is done using a different set of nodes / node ordering than BACKUP). Since the beginning that BACKUP/RESTORE was implemented, we have advised users that they must use a shared data store for the backup files; we recommend S3 for this purpose. Some users use `nodelocal://` but towards a shared filesystem backed via a NAS and iSCSI or similar. *some* users have reported some success with NFS, using fancy options to provide stronger consistency guarantees than defaults.

**Comment by zhangquan913** [comment]
1. Thanks for pointing out that I was dropping a database "under it". Now I always jump out the target db before backup, drop and restore. The "retrieving db name" warning does not occur anymore.
2. I tested with CRDB 2.1.0. Looks the 2.1 file system is more stable. All my full backups, incremental backups and restores worked well so far. I executed all the commands in one pod, and all the backup related files are saved in the same pod. While in 2.0.3, backup files from one command sometimes are found in different pods.
Above is what I found so far. I will test with bigger data and more frequent backups on 2.1.0 to confirm my findings.
